flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +128,293 @@ public DefaultScheduler(
+
+		final Collection<SlotExecutionVertexAssignment> slotExecutionVertexAssignments = allocateSlots(executionVertexDeploymentOptions);
+
+		final Collection<DeploymentHandle> deploymentHandles = createDeploymentHandles(
+			requiredVersionByVertex,
+			deploymentOptionsByVertex,
+			slotExecutionVertexAssignments);
+
+		if (isDeployIndividually()) {
+			deployIndividually(deploymentHandles);
+		} else {
+			waitForAllSlotsAndDeploy(deploymentHandles);
+		}
+	}
+
+	private static Map<ExecutionVertexID, ExecutionVertexDeploymentOption> groupDeploymentOptionsByVertexId(
+			final Collection<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions) {
+		return executionVertexDeploymentOptions.stream().collect(Collectors.toMap(
+				ExecutionVertexDeploymentOption::getExecutionVertexId,
+				Function.identity()));
+	}
+
+	private void prepareToDeployVertices(final Set<ExecutionVertexID> verticesToDeploy) {
+		cancelSlotAssignments(verticesToDeploy);
+		resetForNewExecution(verticesToDeploy);
+		transitionToScheduled(verticesToDeploy);
+	}
+
+	private void cancelSlotAssignments(final Collection<ExecutionVertexID> vertices) {
+		vertices.forEach(executionVertexId -> executionSlotAllocator.cancel(executionVertexId));
+	}
+
+	private Collection<SlotExecutionVertexAssignment> allocateSlots(final Collection<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions) {
+		return executionSlotAllocator.allocateSlotsFor(executionVertexDeploymentOptions
+			.stream()
+			.map(ExecutionVertexDeploymentOption::getExecutionVertexId)
+			.map(this::getExecutionVertex)
+			.map(ExecutionVertexSchedulingRequirementsMapper::from)
+			.collect(Collectors.toList()));
+	}
+
+	private static Collection<DeploymentHandle> createDeploymentHandles(
+		final Map<ExecutionVertexID, ExecutionVertexVersion> requiredVersionByVertex,
+		final Map<ExecutionVertexID, ExecutionVertexDeploymentOption> deploymentOptionsByVertex,
+		final Collection<SlotExecutionVertexAssignment> slotExecutionVertexAssignments) {
+
+		return slotExecutionVertexAssignments
+			.stream()
+			.map(slotExecutionVertexAssignment -> {
+				final ExecutionVertexID executionVertexId = slotExecutionVertexAssignment.getExecutionVertexId();
+				return new DeploymentHandle(
+					requiredVersionByVertex.get(executionVertexId),
+					deploymentOptionsByVertex.get(executionVertexId),
+					slotExecutionVertexAssignment);
+			})
+			.collect(Collectors.toList());
+	}
+
+	private boolean isDeployIndividually() {
+		return schedulingStrategy instanceof LazyFromSourcesSchedulingStrategy;
+	}
+
+	private void deployIndividually(final Collection<DeploymentHandle> deploymentHandles) {
+		for (final DeploymentHandle deploymentHandle : deploymentHandles) {
+			FutureUtils.assertNoException(
+				deploymentHandle
+					.getSlotExecutionVertexAssignment()
+					.getLogicalSlotFuture()
+					.handle(assignResourceOrHandleError(deploymentHandle))
+					.handle(deployOrHandleError(deploymentHandle)));
```

Comment:
Let's add a comment explaining why we need to split `assignResourceOrHandleError` and `deployOrHandleError`.

Reply:
The reason is code reuse. I don't think a comment is needed here. I added a comment to `isDeployIndividually()`: ffc33a982eb3fc6141b6a57f000eaa8f24cf36cb